hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/AbstractJetInstance.java

```diff
@@ -40,6 +53,60 @@ public AbstractJetInstance(HazelcastInstance hazelcastInstance) {
         this.jobRepository = Util.memoizeConcurrent(() -> new JobRepository(this));
     }
 
+    @Nonnull @Override
+    public final Job newJob(@Nonnull DAG dag, @Nonnull JobConfig config) {
+        long jobId = uploadResourcesAndAssignId(config);
+        return newJobProxy(jobId, dag, config);
+    }
+
+    @Nonnull @Override
+    public final Job newJobIfAbsent(@Nonnull DAG dag, @Nonnull JobConfig config) {
+        if (config.getName() == null) {
+            return newJob(dag, config);
+        } else {
+            while (true) {
+                Job job = getJob(config.getName());
+                if (job != null) {
+                    JobStatus status = job.getStatus();
+                    if (status != JobStatus.FAILED && status != JobStatus.COMPLETED) {
+                        return job;
+                    }
+                }
+
+                long jobId = 0;
+                try {
+                    jobId = uploadResourcesAndAssignId(config);
+                    return newJobProxy(jobId, dag, config);
+                } catch (DuplicateActiveJobNameException e) {
+                    logFine(getLogger(), "Could not submit job: %s with duplicate name: %s", idToString(jobId),
```

Comment:
We should sleep about 100ms here, not retry as fast as possible. Maybe even longer